File: sdk/python/openwa/resources/status.py
"""Status (Stories) resource — WhatsApp status updates.

Backed by ``src/modules/status/status.controller.ts``.
NOTE: this is WhatsApp "Status/Stories", distinct from session lifecycle status.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._http import quote_segment
from ..types import SendImageStatusRequest, SendTextStatusRequest, SendVideoStatusRequest, StatusRecord

if TYPE_CHECKING:
    from .._http import HttpExecutor


class StatusResource:
    def __init__(self, http: "HttpExecutor") -> None:
        self._http = http

    def list(self, session_id: str) -> dict[str, list[StatusRecord]]:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/status")

    def from_contact(self, session_id: str, contact_id: str) -> dict[str, list[StatusRecord]]:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/status/{quote_segment(contact_id)}")

    def send_text(self, session_id: str, body: SendTextStatusRequest) -> StatusRecord:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/status/send-text", body=body)

    def send_image(self, session_id: str, body: SendImageStatusRequest) -> StatusRecord:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/status/send-image", body=body)

    def send_video(self, session_id: str, body: SendVideoStatusRequest) -> StatusRecord:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/status/send-video", body=body)

    def delete(self, session_id: str, status_id: str) -> None:
        self._http.request("DELETE", f"/api/sessions/{quote_segment(session_id)}/status/{quote_segment(status_id)}")

File: sdk/javascript/src/resources/status.ts
/**
 * Status (Stories) resource — WhatsApp status updates.
 *
 * Backed by `src/modules/status/status.controller.ts`.
 * NOTE: this is WhatsApp "Status/Stories", distinct from session lifecycle status.
 * @packageDocumentation
 */

import { encodeSegment } from '../http.js';
import type { OpenWAClient } from '../client.js';
import type { SendImageStatusRequest, SendTextStatusRequest, SendVideoStatusRequest, StatusRecord } from '../types.js';

export class StatusResource {
  constructor(private readonly client: OpenWAClient) {}

  /** Get all status updates. */
  list(sessionId: string): Promise<{ statuses: StatusRecord[] }> {
    return this.client.request<{ statuses: StatusRecord[] }>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/status`,
    });
  }

  /** Get status updates from a specific contact. */
  fromContact(sessionId: string, contactId: string): Promise<{ statuses: StatusRecord[] }> {
    return this.client.request<{ statuses: StatusRecord[] }>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/${encodeSegment(contactId)}`,
    });
  }

  /** Post a text status update. */
  sendText(sessionId: string, body: SendTextStatusRequest): Promise<StatusRecord> {
    return this.client.request<StatusRecord>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/send-text`,
      body,
    });
  }

  /** Post an image status update. */
  sendImage(sessionId: string, body: SendImageStatusRequest): Promise<StatusRecord> {
    return this.client.request<StatusRecord>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/send-image`,
      body,
    });
  }

  /** Post a video status update. */
  sendVideo(sessionId: string, body: SendVideoStatusRequest): Promise<StatusRecord> {
    return this.client.request<StatusRecord>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/send-video`,
      body,
    });
  }

  /** Delete a status update by id. */
  delete(sessionId: string, statusId: string): Promise<void> {
    return this.client.request<void>({
      method: 'DELETE',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/${encodeSegment(statusId)}`,
    });
  }
}
